### PR TITLE
Introduce accuracy testing for LLMs in tt-forge.

### PR DESCRIPTION
**Pull request by dgolubovicTT**
Closes tenstorrent/tt-xla#3355 

## Overview: LLM Accuracy Testing with TOP1/TOP5 Metrics                                                                                                                                                                                                                                          
   
  This PR introduces token-level accuracy testing for LLM benchmarks, measuring how well TT device predictions match CPU reference outputs.                                                                                                                                                         
                                  
  **Accuracy Metrics:**
  - **TOP1 Accuracy**: Percentage of tokens where the device's prediction matches the CPU reference model's top prediction (argmax)
  - **TOP5 Accuracy**: Percentage of tokens where the device's prediction appears in the CPU reference model's top-5 predictions

  **Test Corpus:**
  We use "A Tale of Two Cities" by Charles Dickens as our reference text. The text is split into two parts:
  - **Prefill phase** (first half): Input context fed to the model
  - **Decode phase** (second half): Ground truth tokens for validation with teacher forcing

  **Critical Requirement - Sequence Length Matching:**
  The CPU reference outputs (.refpt files) **must** be generated with the same `total_length` as the `input_sequence_length` used during device testing. Mismatched sequence lengths cause accuracy degradation even with teacher forcing, because CPU model and compiled model need to have same input sequence size (context). If you change
  `input_sequence_length` in tests, you must regenerate **all** reference outputs.

NOTE: We currently use almost identical approach to accuracy testing in tt-metal for 1-1 comparison. Same text corpus (Tale of Two Cities) and same top1/top5 metrics. However, since we test accuracy on llms only on n150, we had to use total sequence length 128 (for memory constraints). That means 64 tokens in prefil, and 64 for decode.

### Introduce accuracy testing to test_llms.py
- Add --accuracy-testing argument to test_llms.py tests that we track in benchmark testing.
- Add --batch-size argument to accuracy tests in test_llms.py because we can't fit default batch 32 on device
  due to larger input sequence length required in accuracy testing.
  With batch size 32 and input sequence length, 7B and 8B models failed with OOM issues.
- Run accuracy tests in separate job called: run-n150-accuracy-benchmarks of the perf-benchmark-experimental workflow

### Generating ground truth .refpt files (generate_reference_outputs.py)

  Add generate_reference_outputs.py script that loads Huggingface model, runs it on "Tale of Two Cities" text corpus, and generates a .refpt file containing reference tokens and top-5 predictions for each position.
  Generate reference top1/top5 token predictions for LLM accuracy benchmarking:
  - Loads HuggingFace models on CPU
  - Processes "Tale of Two Cities" text corpus with teacher forcing
  - Outputs .refpt files containing reference tokens and top-k predictions
  - Used by TokenAccuracy class to validate TOP1/TOP5 accuracy during benchmarks
  - Added directory with reference .refpt files - /reference_outputs with README that explains how reference files are    created and used
  - Added `generate_all_reference_outputs.sh` that runs `generate_reference_outputs.py` for a list of models and dumps reference outputs to `benchmark/tt-xla/reference_outputs`

   Ensures reproducibility through eval mode, disabled dropout, greedy decoding,
   and StaticCache matching the benchmark environment. **Reference files must be
   regenerated if input_sequence_length changes.**

   Usage:
      python3 scripts/generate_reference_outputs.py \
        --model "meta-llama/Llama-3.2-1B-Instruct" \
        --output_file "reference_outputs/Llama-3.2-1B-Instruct.refpt" \
        --total_length 128

### Adding shared utility for decode (decode_utils.py)
  Centralize LLM decode operations used by reference output generation and accuracy testing:
  - Teacher forcing generation with ground truth tokens
  - Reference top-k prediction generation for .refpt files
  - Static cache and accuracy testing initialization
  - Logits extraction and top-k token utilities

  **Prevents implementation drift between reference generation and benchmark paths
  by sharing the same decode logic, tokenization, and cache semantics.**

### TokenAccuracy class for validating LLM inference quality(token_accuracy.py)
  - Loads precomputed reference data from .refpt files (tokens, top1/top5 predictions)
  - Validates torch/transformers versions match reference file for reproducibility
  - Splits reference tokens into prefill (input) and decode (ground truth) windows
  - Computes TOP1/TOP5 accuracy by comparing model predictions against reference
  - Provides teacher forcing tokens for deterministic decode loops

### Slight refactoring
  - Simplify static cache initialization using init_static_cache helper
  - Remove unused variables is_multichip and mesh from generate_and_benchmark function.

### CODE REVIEW

File: .github/workflows/perf-bench-matrix.json [hunk 1]
```diff
@@ -281,6 +281,111 @@
...
+      },
+      {
+        "name": "qwen_3_8b_accuracy",
+        "pyreq": "datasets loguru pytest requests tabulate timm torch==2.9.0 torchvision==0.24.0 tqdm transformers==4.57.1",
+        "pytest": "benchmark/tt-xla/test_llms.py::test_qwen_3_8b_accuracy"
+      },
+      {
+        "name": "ministral_8b_accuracy",
+        "pyreq": "datasets loguru pytest requests torch==2.9.0 tqdm transformers==4.57.1",
+        "pytest": "benchmark/tt-xla/test_llms.py::test_ministral_8b_accuracy"
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply .github/workflows/perf-bench-matrix.json]: For start, I really want to track Accuracy on all models that we want to improve and do mixed precision on. I don't think we should jump to reducing reasonable tests before trying to solve it other way.

File: benchmark/tt-xla/llm_benchmark.py [hunk 2]
```diff
@@ -266,14 +280,17 @@ def benchmark_llm_torch_xla(
     shard_spec_fn,
     arch,
     required_pcc,
+    accuracy_testing: bool = False,
+    model_name_for_accuracy: str = None,
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply benchmark/tt-xla/llm_benchmark.py]: Talked with @mvasiljevicTT offline. Her model name comes from pytest test ID. My model name is real Huggingface model name. function that generates these files `generate_reference_outputs` has huggingface model name as attribute and loads it and creates .refpt file with the same name. So it is completely independent on what test name is... Therefore, it would be hacky to force display_name (from Marina's PR).

File: benchmark/tt-xla/llm_benchmark.py [hunk 3]
```diff
@@ -346,27 +384,55 @@ def benchmark_llm_torch_xla(
...
+
+        # Get Tale of Two Cities text from reference data
+        custom_input_prompt = token_accuracy.prepare_ref_tokens(tokenizer)
+        print(
+            f"Using reference text for accuracy testing:"
+            f"\n  Max prefill: {max_prefill} tokens"
+            f"\n  Max decode: {max_decode} tokens"
+            f"\n  Text preview: {custom_input_prompt[:100]}..."
+        )
+
```
Comment (vkovacevicTT) [inline benchmark/tt-xla/llm_benchmark.py]: Maybe put this in an utility function?

File: benchmark/tt-xla/llm_benchmark.py [hunk 4]
```diff
@@ -411,7 +477,8 @@ def benchmark_llm_torch_xla(
     # Warmup run
     print("Warming up...")
     warmup_tokens = min(MIN_STEPS, max_tokens_to_generate)
-    _, _ = generate_and_benchmark(
+    ground_truth_for_warmup = token_accuracy.reference_tokens[:warmup_tokens] if accuracy_testing else None
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply benchmark/tt-xla/llm_benchmark.py]: You are right, we don't

File: benchmark/tt-xla/scripts/generate_all_reference_outputs.sh [hunk 5]
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Batch script to generate .refpt files for all LLM models
+# This is a convenience script and won't be committed to the repo
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply benchmark/tt-xla/scripts/generate_all_reference_outputs.sh]: I think both, because we need this script when we want to regenerate .refpt files...

File: benchmark/tt-xla/llm_benchmark.py [hunk 6]
```diff
@@ -202,6 +225,9 @@ def generate_and_benchmark(
             logits = read_logits_fn(output).to("cpu")
             output_logits.append(logits)
             next_token_ids = logits[:, -1].argmax(dim=-1)
+            predicted_token = next_token_ids[0].item()  # Assuming batch_size=1
+            predicted_tokens.append(predicted_token)
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply benchmark/tt-xla/llm_benchmark.py]: Added support for batch_size != 1 so I think we are good now
